### PR TITLE
Allow trailing slashes in URIs by adding the NormalizePathLayer

### PR DESCRIPTION
**Pull request by spencewenski**
Problem
-------
Some clients may append a trailing slash to URIs, either intentionally or accidentally. We should support both routes that do and don't have trailing slashes for maximum compatibility.

Solution
--------
Normally, when we define a route, it will not have a trailing slash. We can add support for trailing slashes by adding the `NormalizePathLayer`, which will automatically remove the trailing slash from the requested URI, allowing it to map to the route we created without a trailing slash.

### CODE REVIEW

**Comment by kaplanelad** [comment]
@spencewenski can we move this logic to [initializer](https://github.com/loco-rs/loco/tree/master/loco-extras) and add it as a default in our starter. this is allows to user to turn it off if they don't want it

**Comment by spencewenski** [comment]
Hi @kaplanelad, thanks for the suggestion! I got a prototype of the initializer approach working. Below is a snippet of the code. My only concern is I'm not sure if there's any implications to routing when using this approach. I'll read some docs and play around with this some more and if everything seems okay I'll update the PR with the initializer approach.

Code snippet:
```rust
async fn after_routes(&self, router: Router, _ctx: &AppContext) -> Result<Router> {
    let router = NormalizePathLayer::trim_trailing_slash().layer(router);
    let router = Router::new().nest_service("", router);
    Ok(router)
}
```

**Comment by kaplanelad** [comment]
You can follow loco-extras [here](https://github.com/loco-rs/loco/blob/master/loco-extras/src/initializers/mod.rs).
The initializer gets `AxumRouter`. 

[Here ](https://github.com/loco-rs/loco/blob/master/examples/demo/src/app.rs#L53) you can see use initializer with `prometheus`. 

i'm suggestion to do the same.

**Comment by spencewenski** [comment]
Yup, I understand. I updated the PR. My concerns were related to the behavior of `nest_service` -- specifically that the [nested routes have the prefix stripped](https://docs.rs/axum/latest/axum/struct.Router.html#how-the-uri-changes). However, since we're nesting at the root route, this doesn't appear to be a problem.

File: Cargo.toml [hunk 1]
```diff
@@ -131,6 +122,18 @@ object_store = { version = "0.9.0", default-features = false }
...
+tower-http = { version = "0.5.0", features = [
+  "trace",
+  "catch-panic",
+  "timeout",
+  "add-extension",
+  "cors",
+  "fs",
+  "set-header",
+  "compression-full",
+  "normalize-path"
```
Comment (kaplanelad) [inline Cargo.toml:135]: This is not a requirement from loco framework, this means that the `normalize-path` will includded for all the users even if they not using your initializer. can you please move it to the loco-extras cargo file?

File: examples/demo/src/app.rs [hunk 2]
```diff
@@ -46,6 +47,7 @@ impl Hooks for App {
             Box::new(initializers::axum_session::AxumSessionInitializer),
             Box::new(initializers::view_engine::ViewEngineInitializer),
             Box::new(initializers::hello_view_engine::HelloViewEngineInitializer),
+            Box::new(loco_extras::initializers::normalize_path::NormalizePathInitializer),
```
Comment (kaplanelad) [inline examples/demo/src/app.rs:50]: Can you please add a test case in `https://github.com/loco-rs/loco/tree/master/examples/demo/tests/requests`?

**Review by kaplanelad — CHANGES_REQUESTED** [review]
Thank you, @spencewenski! This is exactly what I was looking for.
Could you please take a look to review two small notes?

File: examples/demo/src/app.rs (hunk 2 above)
Comment (spencewenski) [reply examples/demo/src/app.rs:49]: Added tests using the `ping` endpoint.

File: Cargo.toml (hunk 1 above)
Comment (spencewenski) [reply Cargo.toml]: Done. Moved to the `loco-extras` cargo file.

**Review by kaplanelad — APPROVED** [review]

**Comment by spencewenski** [comment]
Hey, @kaplanelad , thanks approving the PR! Are you able to merge it for me? I don’t have push permissions.

**Comment by jondot** [comment]
@kaplanelad I’m wondering is this useful as a default inside loco for all routes?

**Comment by jondot** [comment]
I mean why not include it not as an initializer but as an integrated later in all routers instead of our custom code for handing ‘//‘?

**Pull request merged**